Support mlaunch on windows

zhaoyi0113 opened the pull request:
Add support on running mlaunch in windows.

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -668,12 +671,12 @@ def kill(self):
             if port in matches:
                 p = processes[port]
                 p.send_signal(sig)
+                print " %s on port %i, pid=%i" % (p.name, port, p.pid)
```
This line is also printed under the `--verbose` argument.

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -1422,7 +1440,11 @@ def _construct_mongos(self, logpath, port, configdb):
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	else:
+            command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), newLogPath, port, configdb, auth_param, extra)
```
I believe `newLogPath` is not defined when the OS is not Windows, resulting in the error `UnboundLocalError: local variable 'newLogPath' referenced before assignment`

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -1422,7 +1440,11 @@ def _construct_mongos(self, logpath, port, configdb):
-        command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	if os.name == 'nt':
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
```
I assume `newLogPath` was meant to be used in this line?

zhaoyi0113 commented:
Thaks for testing it. Below is my comments regard to your test cases:

1. I haven't tried that paratemers yet. I will let you know after I investigate it.

2. You need to install `openssl` on windows in order to make `--auth` work

3. This command works pretty well for me, below is the output:

```
>mlaunch init --sharded 3 --replicaset --config 3 --mongos 2
launching: start on port 27019
launching: start on port 27020
launching: start on port 27021
launching: start on port 27022
launching: start on port 27023
launching: start on port 27024
launching: start on port 27025
launching: start on port 27026
launching: start on port 27027
launching: config server on port 27028
launching: config server on port 27029
launching: config server on port 27030
replica set 'configRepl' initialized.
replica set 'shard01' initialized.
replica set 'shard02' initialized.
replica set 'shard03' initialized.
launching: start on port 27017
launching: start on port 27018
adding shards. can take up to 30 seconds...
```
4. I will remove the log for this case.

zhaoyi0113 commented:
I see there is an error on travis ci. Could you let me know what os it is running?

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -1422,7 +1439,11 @@ def _construct_mongos(self, logpath, port, configdb):
 
         path = self.args['binarypath'] or ''
-        command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	      if os.name == 'nt':
```
There's a mix of tabs & spaces in this line

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -1422,7 +1439,11 @@ def _construct_mongos(self, logpath, port, configdb):
+	      if os.name == 'nt':
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), newLogPath, port, configdb, auth_param, extra)
+	      else:
```
There's a mix of tabs & spaces in this line

kevinadi commented:
The Travis build was using Ubuntu 12.04.5 LTS. However, the failures were due to a mix of tabs & spaces in lines 1442 and 1445.

zhaoyi0113 commented:
Ok, thanks for pointing out. I have fixed the error and pushed. Could you let me know what else need to be done in order to merge in this PR?

kevinadi commented:
@zhaoyi0113 thanks for the update. Sorry it took some time to review this.

I have found one last thing to update. Apparently extra parameters are not passed on to `mlaunch` on Windows. For example:

```
mlaunch init --single --storageEngine mmapv1
```

The `--storageEngine` parameter is not passed on in Windows, but is on OSX.

I tracked the issue to the `_filter_valid_arguments()` function, specifically in line 1062.

Do you mind updating the PR to accommodate extra parameters? Thanks.

kevinadi commented on mtools/mlaunch/mlaunch.py:
```diff
@@ -1128,7 +1130,13 @@ def _start_on_ports(self, ports, wait=False, overrideAuth=False):
             try:
-                ret = subprocess.check_output([command_str], stderr=subprocess.STDOUT, shell=True)
+                if os.name == 'nt':
+                    # print "command str ", filter(None, command_str.split(' ')), "."
```
Could you remove this commented line please? I believe it's a development debug line, so we don't need it anymore. Thanks.

kevinadi commented:
Also I have put up a comment regarding the development debug line (line 1134), I think we don't need that line anymore 👍

zhaoyi0113 commented:
I have fixed the `mlaunch init --single --storageEngine mmapv1` issue and removed the comment code. Please let me know what else need to be done for this PR. Thanks.

stennie commented:
@zhaoyi0113 Great stuff .. thanks for your contribution!

Regards,
Stennie
